Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,197 @@
+                }
+                self.type = .ipLiteral
+            } else {
+                let parts = hostAndPort.components(separatedBy: ":")
+                host = String(parts[0])
+                guard parts.count <= 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "expected only a single port delimiter ':' in \(hostAndPort)"
+                    )
+                }
+                if parts.count > 1 {
+                    port = try HostIdentifier.verifyPort(from: parts[1], scheme: scheme)
+                } else {
+                    port = scheme == .srv ? nil : 27017
+                }
+                self.type = .hostname
+            }
+            self.host = host
+            self.port = port
+        }
+
+        public var description: String {
+            switch self.type {
+            case .ipLiteral:
+                if let prt = port {
+                    return "[\(self.host)]:\(prt)"
+                } else {
+                    return "[\(self.host)]"
+                }
+            default:
+                if let prt = port {
+                    return "\(self.host):\(prt)"
+                } else {
+                    return "\(self.host)"
+                }
+            }
+        }
+    }
+
+    /// Parses a new `MongoConnectionString` instance from the provided string.
+    /// - Throws:
+    ///   - `MongoError.InvalidArgumentError` if the input is invalid.
+    public init(throwsIfInvalid input: String) throws {
+        let schemeAndRest = input.components(separatedBy: "://")
+        guard schemeAndRest.count == 2, let scheme = Scheme(schemeAndRest[0]) else {
+            throw MongoError.InvalidArgumentError(
+                message: "Invalid connection string scheme, expecting \'mongodb\' or \'mongodb+srv\'"
+            )
+        }
+        guard !schemeAndRest[1].isEmpty else {
+            throw MongoError.InvalidArgumentError(message: "Invalid connection string")
+        }
+        let identifiersAndOptions = schemeAndRest[1].components(separatedBy: "/")
+        guard identifiersAndOptions.count <= 2 else {
+            throw MongoError.InvalidArgumentError(message: "Host with unescaped slash")
+        }
+        let userAndHost = identifiersAndOptions[0].components(separatedBy: "@")
+        guard userAndHost.count <= 2 else {
+            throw MongoError.InvalidArgumentError(message: "Invalid user information")
+        }
+        let userInfo = userAndHost.count == 2 ? userAndHost[0].components(separatedBy: ":") : nil
+        let hostString = userInfo != nil ? userAndHost[1] : userAndHost[0]
+        let hosts = try hostString.components(separatedBy: ",").map {
+            try HostIdentifier($0, scheme: scheme)
+        }
+        if case .srv = scheme {
+            guard hosts.count == 1 else {
+                throw MongoError.InvalidArgumentError(message: "mongodb+srv with multiple service names")
```

Comment:
Ditto updating this message to be more specific. Also, I don't think "service names" is the right term here (I think the description for the test in `invalid-uris.yml` is incorrect). Something like `"only a single host identifier may be specified in a mongodb+srv connection string"` seems good.

Reply:
done